### PR TITLE
[core][state][job] Supporting job listing(getting) and logs from state API

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
This PR adds better support for listing jobs (`ray list jobs`) and getting submission job logs (`ray logs job --id`) 

The state API client mirrors the implementation of job endpoint and depends on job's implementations for retrieving job related info. 

[PR] Output of running 2 submission jobs + 1 driver job: 
image

[Master] `ray list jobs`
- Currently, driver jobs are not included in the output. 
image

[Master] `ray job list` 
- The job CLI is a bit hard to parse the results, but it includes the driver job. 
image

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/state/state_cli.py
Comment (rickyyx) [inline python/ray/experimental/state/state_cli.py:311]: we need to control the to dict routine here since pydantic and python dataclasses don't really work the same with `dataclasses.asdict`.

File: dashboard/modules/job/pydantic_models.py
Comment (architkulkarni) [inline dashboard/modules/job/pydantic_models.py:42]: Why remove this?

**Review by architkulkarni — APPROVED** [review]
Looks good to me.  Do you mind posting a comparison of the output before/after this PR?

Nit: Is the public filter "DRIVER" vs "SUBMISSION" already GA or are we allowed to change it? The reason is we formally changed "Ray Job Submission" to "Ray Jobs API" in the docs.  So maybe we can change "SUBMISSION" to "JOB_API", cc @scottsun94 for thoughts.

File: dashboard/modules/job/pydantic_models.py
Comment (rickyyx) [reply dashboard/modules/job/pydantic_models.py:42]: it's duplicate with https://github.com/ray-project/ray/pull/35124/files/e231a5ca7141af307f3cefd16cbb31e48789f134#diff-d1972d6e703f3aa016325518da046bfa3647810e6315e8bd351f95912a02d8a8L66

Comment (architkulkarni) [reply dashboard/modules/job/pydantic_models.py:42]: Ah good catch

**Comment by scottsun94** [comment]
I don't think we publicly expose the two job types before. Can we align on this first? 
cc: @rkooo567 @alanwguo @akshay-anyscale any other options or objections?

**Comment by rickyyx** [comment]
> Looks good to me. Do you mind posting a comparison of the output before/after this PR?
> 
> Nit: Is the public filter "DRIVER" vs "SUBMISSION" already GA or are we allowed to change it? The reason is we formally changed "Ray Job Submission" to "Ray Jobs API" in the docs. So maybe we can change "SUBMISSION" to "JOB_API", cc @scottsun94 for thoughts.

Is not GA - and this is kind of outside of the PR also. The specific type name of the jobs are orthogonal to this PR I guess?

File: dashboard/modules/job/common.py
Comment (rkooo567) [inline dashboard/modules/job/common.py:28]: Use  submission id instead since it is technically a submission id? 
```suggestion
JOB_LOGS_PATH_TEMPLATE = "job-driver-{submission_id}.log"
```

File: dashboard/modules/log/log_manager.py
Comment (rkooo567) [inline dashboard/modules/log/log_manager.py:323]: is it job id or submission id?

File: dashboard/state_aggregator.py
Comment (rkooo567) [inline dashboard/state_aggregator.py:88]: isn't it asdict?

File: python/ray/experimental/state/api.py
Comment (rkooo567) [inline python/ray/experimental/state/api.py:1210]: submission_id?

File: python/ray/experimental/state/state_cli.py
Comment (rkooo567) [reply python/ray/experimental/state/state_cli.py:309]: I guess we should eventually use pydantic only. but sgtm. (cc @alanwguo )

File: python/ray/experimental/state/state_cli.py
Comment (rkooo567) [inline python/ray/experimental/state/state_cli.py:816]: submission_id?

File: python/ray/experimental/state/state_cli.py
Comment (rkooo567) [inline python/ray/experimental/state/state_cli.py:1180]: submission_id?

File: python/ray/experimental/state/state_manager.py
Comment (rkooo567) [inline python/ray/experimental/state/state_manager.py:341]: can you comment on list_jobs that logic should be unified?

**Review by rkooo567 — APPROVED** [review]
@scottsun94 we should probably use the term submission_id? That's the word exposed from the dashboard UI, so using job id here would be confusing. 

Also, LGTM in general (it's pretty clean :)). I only have comments regarding job_id -> submission_id. If anyone wants to discuss it, please create a sync up meeting.

File: dashboard/state_aggregator.py
Comment (rickyyx) [reply dashboard/state_aggregator.py:88]: no, it's mapping to field -> field type, rather than values.

**Comment by rickyyx** [comment]
- Using submission id in the state API
If we are changing the phrase, I think we could do it together when the Job API changes the wording of the submission id.

File: python/ray/experimental/state/state_cli.py
Comment (rkooo567) [inline python/ray/experimental/state/state_cli.py:1185]: Isn't it supposed to be --submission-id?

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
I need final approval from @scottsun94 before merging this PR

**Comment by rkooo567** [comment]
@rickyyx cc for the test failures (can you verify if it is related?)

**Comment by rickyyx** [comment]
I think it's related.

**Comment by scottsun94** [comment]
> * Using submission id in the state API
>   If we are changing the phrase, I think we could do it together when the Job API changes the wording of the submission id.

Let's keep the status quo fro now and discuss the naming change separately. Can we indicate somewhere that the job type field is still experimental and might change? (at least in the doc) @rickyyx

**Comment by scottsun94** [comment]
LGTM. Thanks!

**Pull request merged**
